In [1]:
import tensorflow as tf

tf.config.list_physical_devices()

2023-11-06 14:07:09.143116: I tensorflow/core/util/port.cc:113] oneDNN custom operations are on. You may see slightly different numerical results due to floating-point round-off errors from different computation orders. To turn them off, set the environment variable `TF_ENABLE_ONEDNN_OPTS=0`.
2023-11-06 14:07:09.164098: E external/local_xla/xla/stream_executor/cuda/cuda_dnn.cc:9261] Unable to register cuDNN factory: Attempting to register factory for plugin cuDNN when one has already been registered
2023-11-06 14:07:09.164114: E external/local_xla/xla/stream_executor/cuda/cuda_fft.cc:607] Unable to register cuFFT factory: Attempting to register factory for plugin cuFFT when one has already been registered
2023-11-06 14:07:09.164672: E external/local_xla/xla/stream_executor/cuda/cuda_blas.cc:1515] Unable to register cuBLAS factory: Attempting to register factory for plugin cuBLAS when one has already been registered
2023-11-06 14:07:09.168332: I tensorflow/core/platform/cpu_feature_guar

[PhysicalDevice(name='/physical_device:CPU:0', device_type='CPU'),
 PhysicalDevice(name='/physical_device:GPU:0', device_type='GPU')]

In [3]:
tf.version.VERSION

'2.15.0-rc1'

In [1]:
import torch
torch.cuda.is_available()

True

## Try out DeepFace

In [1]:
# use this to disable GPU for timeit experiments
#import os
#os.environ["CUDA_VISIBLE_DEVICES"] = "-1"

In [2]:
from deepface import DeepFace

2023-11-06 16:53:53.458386: I tensorflow/core/util/port.cc:113] oneDNN custom operations are on. You may see slightly different numerical results due to floating-point round-off errors from different computation orders. To turn them off, set the environment variable `TF_ENABLE_ONEDNN_OPTS=0`.
2023-11-06 16:53:53.479395: E external/local_xla/xla/stream_executor/cuda/cuda_dnn.cc:9261] Unable to register cuDNN factory: Attempting to register factory for plugin cuDNN when one has already been registered
2023-11-06 16:53:53.479411: E external/local_xla/xla/stream_executor/cuda/cuda_fft.cc:607] Unable to register cuFFT factory: Attempting to register factory for plugin cuFFT when one has already been registered
2023-11-06 16:53:53.479979: E external/local_xla/xla/stream_executor/cuda/cuda_blas.cc:1515] Unable to register cuBLAS factory: Attempting to register factory for plugin cuBLAS when one has already been registered
2023-11-06 16:53:53.483543: I tensorflow/core/platform/cpu_feature_guar

In [3]:
DeepFace.analyze(img_path='./data/wiki/00/23300_1962-06-19_2011.jpg', actions=['age', 'gender', 'race', 'emotion'])

2023-11-06 16:54:00.034946: E external/local_xla/xla/stream_executor/cuda/cuda_driver.cc:274] failed call to cuInit: CUDA_ERROR_NO_DEVICE: no CUDA-capable device is detected
2023-11-06 16:54:00.034968: I external/local_xla/xla/stream_executor/cuda/cuda_diagnostics.cc:129] retrieving CUDA diagnostic information for host: workstation-322
2023-11-06 16:54:00.034971: I external/local_xla/xla/stream_executor/cuda/cuda_diagnostics.cc:136] hostname: workstation-322
2023-11-06 16:54:00.035039: I external/local_xla/xla/stream_executor/cuda/cuda_diagnostics.cc:159] libcuda reported version is: 535.129.3
2023-11-06 16:54:00.035049: I external/local_xla/xla/stream_executor/cuda/cuda_diagnostics.cc:163] kernel reported version is: 535.129.3
2023-11-06 16:54:00.035051: I external/local_xla/xla/stream_executor/cuda/cuda_diagnostics.cc:241] kernel version seems to match DSO: 535.129.3
Action: emotion: 100%|██████████| 4/4 [00:00<00:00,  9.82it/s]


[{'age': 24,
  'region': {'x': 103, 'y': 86, 'w': 173, 'h': 173},
  'gender': {'Woman': 1.211789809167385, 'Man': 98.7882137298584},
  'dominant_gender': 'Man',
  'race': {'asian': 1.7656517301486491e-06,
   'indian': 3.938198890807646e-07,
   'black': 9.035110980465474e-11,
   'white': 99.97030496420294,
   'middle eastern': 0.010561766150363753,
   'latino hispanic': 0.019136037421139326},
  'dominant_race': 'white',
  'emotion': {'angry': 0.012012210959742774,
   'disgust': 1.5504772265598492e-10,
   'fear': 0.00785635908892009,
   'happy': 1.9288820709130678,
   'sad': 12.62752561241083,
   'surprise': 6.748335751763886e-07,
   'neutral': 85.42372162059287},
  'dominant_emotion': 'neutral'},
 {'age': 33,
  'region': {'x': 36, 'y': 81, 'w': 189, 'h': 189},
  'gender': {'Woman': 99.76291060447693, 'Man': 0.2370966598391533},
  'dominant_gender': 'Woman',
  'race': {'asian': 0.03584688238333911,
   'indian': 0.018068469944410026,
   'black': 0.0008432944014202803,
   'white': 92.74369

In [6]:
%%timeit

DeepFace.analyze(img_path='./data/wiki/00/23300_1962-06-19_2011.jpg', actions=['age', 'gender', 'race', 'emotion'])

Action: emotion: 100%|██████████| 4/4 [00:00<00:00,  9.36it/s]

885 ms ± 27.5 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)


In [138]:
from deepface.DeepFace import *

def analyze_quicker(
    images,
    actions=("emotion", "age", "gender", "race"),
    enforce_detection=True,
    detector_backend="opencv",
    align=True,
    silent=False,
):

    # ---------------------------------
    # validate actions
    if isinstance(actions, str):
        actions = (actions,)

    actions = list(actions)
    # ---------------------------------
    # build models
    models = {}
    #if "emotion" in actions:
    #    models["emotion"] = build_model("Emotion")

    if "age" in actions:
        models["age"] = build_model("Age")

    if "gender" in actions:
        models["gender"] = build_model("Gender")

    if "race" in actions:
        models["race"] = build_model("Race")
    # ---------------------------------

    image_df = pd.DataFrame()

    for img_path in images:
        img_objs = functions.extract_faces(
            img=img_path,
            target_size=(224, 224),
            detector_backend=detector_backend,
            grayscale=False,
            enforce_detection=enforce_detection,
            align=align,
        )
        tmp_df = pd.DataFrame.from_records(img_objs, columns=['img_content', 'img_region', 'confidence'])
        tmp_df['img_path'] = img_path

        image_df = pd.concat([image_df, tmp_df]).reset_index(drop=True)

    batch_iter = 0
    batch_step = 100

    results_df = pd.DataFrame()
    
    while batch_iter < len(image_df):

        batch_df = image_df.iloc[batch_iter:batch_iter+batch_step]

        if len(batch_df) > 0:
            tmp_results_df = batch_df[['img_path', 'img_region', 'confidence']].reset_index(drop=True)
            batch_content = np.stack(batch_df.img_content)[:,0,:,:]  # results in batch size * width * height * RGB

            gender_prediction = models["gender"].predict(batch_content, verbose=0)
            gender_df = pd.DataFrame(gender_prediction, columns=Gender.labels)

            age_prediction = models["age"].predict(batch_content, verbose=0)
            age_df = pd.DataFrame(age_prediction, columns=np.arange(0,101))
            
            race_prediction = models["race"].predict(batch_content, verbose=0)
            race_df = pd.DataFrame(race_prediction, columns=Race.labels)

            tmp_results_df = pd.concat([tmp_results_df, gender_df, race_df, age_df], axis=1)

            results_df = pd.concat([results_df, tmp_results_df])
        batch_iter += batch_step
    
    return results_df.reset_index(drop=True)

In [139]:
pred_results = analyze_quicker(images=['./data/wiki/00/23300_1962-06-19_2011.jpg']*100, actions=['age', 'gender', 'race'], enforce_detection=False)

In [140]:
pred_results

img_path  \
0    ./data/wiki/00/23300_1962-06-19_2011.jpg   
1    ./data/wiki/00/23300_1962-06-19_2011.jpg   
2    ./data/wiki/00/23300_1962-06-19_2011.jpg   
3    ./data/wiki/00/23300_1962-06-19_2011.jpg   
4    ./data/wiki/00/23300_1962-06-19_2011.jpg   
..                                        ...   
195  ./data/wiki/00/23300_1962-06-19_2011.jpg   
196  ./data/wiki/00/23300_1962-06-19_2011.jpg   
197  ./data/wiki/00/23300_1962-06-19_2011.jpg   
198  ./data/wiki/00/23300_1962-06-19_2011.jpg   
199  ./data/wiki/00/23300_1962-06-19_2011.jpg   

                                  img_region  confidence     Woman       Man  \
0    {'x': 103, 'y': 86, 'w': 173, 'h': 173}    4.070868  0.012118  0.987882   
1     {'x': 36, 'y': 81, 'w': 189, 'h': 189}    6.651494  0.997629  0.002371   
2    {'x': 103, 'y': 86, 'w': 173, 'h': 173}    4.070868  0.012118  0.987882   
3     {'x': 36, 'y': 81, 'w': 189, 'h': 189}    6.651494  0.997629  0.002371   
4    {'x': 103, 'y': 86, 'w': 173, 'h': 173}    4.070868  0.012118  0.987882   
..                                       ...         ...       ...       ...   
195   {'x': 36, 'y': 81, 'w': 189, 'h': 189}    6.651494  0.997629  0.002371   
196  {'x': 103, 'y': 86, 'w': 173, 'h': 173}    4.070868  0.012118  0.987882   
197   {'x': 36, 'y': 81, 'w': 189, 'h': 189}    6.651494  0.997627  0.002373   
198  {'x': 103, 'y': 86, 'w': 173, 'h': 173}    4.070868  0.012118  0.987882   
199   {'x': 36, 'y': 81, 'w': 189, 'h': 189}    6.651494  0.997627  0.002373   

            asian        indian         black     white  middle eastern  ...  \
0    1.765571e-08  3.937989e-09  9.034558e-13  0.999703        0.000106  ...   
1    3.584693e-04  1.806844e-04  8.432927e-06  0.927437        0.016947  ...   
2    1.765571e-08  3.937989e-09  9.034558e-13  0.999703        0.000106  ...   
3    3.584693e-04  1.806844e-04  8.432927e-06  0.927437        0.016947  ...   
4    1.765571e-08  3.937989e-09  9.034558e-13  0.999703        0.000106  ...   
..            ...           ...           ...       ...             ...  ...   
195  3.584693e-04  1.806844e-04  8.432927e-06  0.927437        0.016947  ...   
196  1.766073e-08  3.936232e-09  9.031890e-13  0.999703        0.000106  ...   
197  3.584900e-04  1.806460e-04  8.430736e-06  0.927456        0.016943  ...   
198  1.766073e-08  3.936232e-09  9.031890e-13  0.999703        0.000106  ...   
199  3.584900e-04  1.806460e-04  8.430736e-06  0.927456        0.016943  ...   

               91            92            93            94            95  \
0    0.000000e+00  0.000000e+00  0.000000e+00  0.000000e+00  0.000000e+00   
1    1.223829e-11  9.470796e-15  9.501270e-14  4.138538e-13  2.901136e-14   
2    0.000000e+00  0.000000e+00  0.000000e+00  0.000000e+00  0.000000e+00   
3    1.223829e-11  9.470796e-15  9.501270e-14  4.138538e-13  2.901136e-14   
4    0.000000e+00  0.000000e+00  0.000000e+00  0.000000e+00  0.000000e+00   
..            ...           ...           ...           ...           ...   
195  1.223829e-11  9.470796e-15  9.501270e-14  4.138538e-13  2.901136e-14   
196  0.000000e+00  0.000000e+00  0.000000e+00  0.000000e+00  0.000000e+00   
197  1.227010e-11  9.497890e-15  9.535778e-14  4.157263e-13  2.908697e-14   
198  0.000000e+00  0.000000e+00  0.000000e+00  0.000000e+00  0.000000e+00   
199  1.227010e-11  9.497890e-15  9.535778e-14  4.157263e-13  2.908697e-14   

               96            97            98            99           100  
0    0.000000e+00  0.000000e+00  0.000000e+00  0.000000e+00  0.000000e+00  
1    4.385113e-13  5.492407e-13  1.042487e-10  5.650783e-12  3.433942e-09  
2    0.000000e+00  0.000000e+00  0.000000e+00  0.000000e+00  0.000000e+00  
3    4.385113e-13  5.492407e-13  1.042487e-10  5.650783e-12  3.433942e-09  
4    0.000000e+00  0.000000e+00  0.000000e+00  0.000000e+00  0.000000e+00  
..            ...           ...           ...           ...           ...  
195  4.385113e-13  5.492407e-13  1.042487e-10  5.650783e-12  3.433942

### DeepFace Optimization

(using an image where 2 faces are detected)

#### Using CPU
**predict ALL (race, gender, age) attributes**
- With 1 prediction: 900 ms

#### Using GPU

**predict ONLY race**
- Without prediction: 28 ms
- With single prediction: 85 ms
- With 10 predictions: 117 ms
- With 100 predictions: 500 ms

**predict ALL (race, gender, age) attributes**
- With 1 prediction: 196 ms
- With 100 predictions: 1380 ms

In [10]:
print("CPU:", round(523051 * 0.9 / 60 / 60,2), "h (", round(523051 * 0.9 / 60 / 60 / 24, 2), "d)")  # images -> seconds -> minutes -> hours
print("GPU, without batches:", round(523051 * 0.2 / 60 / 60, 2), "h")  # images -> seconds -> minutes -> hours
print("GPU, with batches:", round(523051 / 100 * 1.4 / 60 / 60, 2), "h")  # images -> batches -> seconds -> minutes -> hours

CPU: 130.76 h ( 5.45 d)
GPU, without batches: 29.06 h
GPU, with batches: 2.03 h


## Try out FairFace

In [2]:
import scipy.io
import numpy as np
import pandas as pd

In [2]:
# Load MATLAB files
wiki = scipy.io.loadmat('./data/imdb-wiki/wiki.mat')
imdb = scipy.io.loadmat('./data/imdb-wiki/imdb.mat')


In [3]:
full_path_wiki = np.stack(wiki['wiki']['full_path'][0][0][0]).flatten()
full_path_imdb = np.stack(imdb['imdb']['full_path'][0][0][0]).flatten()
full_path_imdb

array(['01/nm0000001_rm124825600_1899-5-10_1968.jpg',
       '01/nm0000001_rm3343756032_1899-5-10_1970.jpg',
       '01/nm0000001_rm577153792_1899-5-10_1968.jpg', ...,
       '08/nm3994408_rm926592512_1989-12-29_2011.jpg',
       '08/nm3994408_rm943369728_1989-12-29_2011.jpg',
       '08/nm3994408_rm976924160_1989-12-29_2011.jpg'], dtype='<U45')

In [4]:
len(full_path_imdb) + len(full_path_wiki)

523051

In [6]:
test_df = pd.DataFrame(full_path_imdb[:100], columns=['img_path'])
test_df['img_path'] = '../fairpair/data/imdb_crop/' + test_df['img_path']
test_df

img_path
0   ../fairpair/data/imdb_crop/01/nm0000001_rm1248...
1   ../fairpair/data/imdb_crop/01/nm0000001_rm3343...
2   ../fairpair/data/imdb_crop/01/nm0000001_rm5771...
3   ../fairpair/data/imdb_crop/01/nm0000001_rm9469...
4   ../fairpair/data/imdb_crop/01/nm0000001_rm9804...
..                                                ...
95  ../fairpair/data/imdb_crop/06/nm0000006_rm4692...
96  ../fairpair/data/imdb_crop/08/nm0000008_rm1221...
97  ../fairpair/data/imdb_crop/08/nm0000008_rm1238...
98  ../fairpair/data/imdb_crop/08/nm0000008_rm1271...
99  ../fairpair/data/imdb_crop/08/nm0000008_rm1288...

[100 rows x 1 columns]

In [7]:
test_df.to_csv('./data/imdb-wiki/fairface-test.csv')

## Load processed images

In [3]:
labeled_df = pd.read_csv('./data/annotated_results/annotated_pictures_0.csv')
labeled_df

img_path  \
0   ./data/wiki_crop/17/10000217_1981-05-05_2009.jpg   
1   ./data/wiki_crop/48/10000548_1925-04-04_1964.jpg   
2     ./data/wiki_crop/12/100012_1948-07-03_2008.jpg   
3   ./data/wiki_crop/65/10001965_1930-05-23_1961.jpg   
4   ./data/wiki_crop/16/10002116_1971-05-31_2012.jpg   
..                                               ...   
95   ./data/wiki_crop/40/1004540_1975-08-25_2011.jpg   
96   ./data/wiki_crop/04/1004604_1981-11-27_2012.jpg   
97  ./data/wiki_crop/06/10046706_1933-04-21_1962.jpg   
98  ./data/wiki_crop/56/10046756_1987-07-16_2014.jpg   
99  ./data/wiki_crop/87/10047987_1940-05-15_1969.jpg   

                                  img_region  confidence     Woman       Man  \
0     {'x': 85, 'y': 50, 'w': 122, 'h': 122}    3.814839  0.004158  0.995842   
1       {'x': 0, 'y': 0, 'w': 186, 'h': 186}    0.000000  0.003327  0.996673   
2    {'x': 113, 'y': 75, 'w': 286, 'h': 286}    7.710119  0.000102  0.999898   
3       {'x': 0, 'y': 0, 'w': 500, 'h': 347}    0.000000  0.089901  0.910099   
4       {'x': 0, 'y': 0, 'w': 173, 'h': 173}    0.000000  0.999984  0.000016   
..                                       ...         ...       ...       ...   
95   {'x': 119, 'y': 84, 'w': 224, 'h': 224}    6.953803  0.999970  0.000030   
96      {'x': 0, 'y': 0, 'w': 149, 'h': 149}    0.000000  0.005817  0.994183   
97      {'x': 24, 'y': 21, 'w': 39, 'h': 39}    8.620349  0.000022  0.999978   
98    {'x': 37, 'y': 39, 'w': 206, 'h': 206}    3.802662  0.963850  0.036150   
99  {'x': 162, 'y': 148, 'w': 130, 'h': 130}    5.080689  0.000097  0.999903   

           asian        indian         black     white  middle eastern  ...  \
0   4.853634e-02  5.958896e-02  2.153041e-02  0.355860    2.303293e-01  ...   
1   7.334924e-04  1.213157e-03  1.549434e-04  0.903966    7.414950e-02  ...   
2   4.954240e-02  1.942875e-01  5.793767e-02  0.162475    3.176293e-01  ...   
3   1.003605e-02  5.195348e-05  2.168058e-05  0.955353    1.792027e-02  ...   
4   3.139276e-09  2.778450e-09  1.213339e-12  0.999900    5.431611e-05  ...   
..           ...           ...           ...       ...             ...  ...   
95  4.664802e-13  1.483310e-13  2.653851e-17  1.000000    2.250399e-07  ...   
96  1.892886e-03  2.371577e-02  1.180303e-01  0.465310    2.961581e-01  ...   
97  5.013351e-06  1.577920e-05  2.168542e-07  0.967106    1.803476e-02  ...   
98  1.217312e-03  5.659866e-04  5.423994e-05  0.879987    4.835310e-02  ...   
99  7.796633e-05  6.766038e-04  1.686563e-05  0.846467    1.044902e-01  ...   

              91            92            93            94            95  \
0   7.788205e-32  1.230157e-36  2.912823e-35  4.639201e-36  8.101781e-34   
1   5.732801e-08  3.169249e-10  2.491972e-08  3.304694e-06  1.213526e-10   
2   5.015985e-14  3.538318e-18  5.286121e-16  9.664587e-13  3.717189e-19   
3   1.083414e-09  1.809404e-12  1.336072e-10  8.004656e-09  1.420841e-12   
4   1.235424e-07  5.599671e-09  1.814577e-08  1.185982e-07  1.091720e-08   
..           ...           ...           ...           ...           ...   
95  1.724695e-15  2.451515e-18  2.186962e-18  2.458408e-18  6.363941e-18   
96  1.372630e-11  3.905264e-15  7.593836e-13  5.156936e-11  3.146095e-15   
97  4.130149e-28  1.048550e-32  1.778765e-32  2.794496e-33  3.196573e-30   
98  3.222135e-12  2.038118e-16  2.337376e-14  2.243559e-13  1.817827e-15   
99  4.165187e-11  1.971300e-15  8.813369e-13  1.205335e-11  9.039848e-15   

              96            97            98            99           100  
0   1.435297e-32  7.651610e-28  6.354907e-22  1.301743e-25  2.544115e-32  
1   1.694896e-09  5.908470e-10  7.823120e-09  6.609567e-10  2.090870e-05  
2   3.405330e-16  7.258854e-15  1.943376e-12  1.333881e-14  9.771958e-11  
3   2.637161e-11  1.826834e-11  1.513905e-09  7.955166e-11  5.761927e-07  
4   1.494612e-08  1.377759e-09  2.224946e-08  5.173749e-09  1.559782e-06  
..           ...           ...           ...           ...           ...  
95  3.127070e-16  1.752